p2panda-sync-next/src/log_sync.rs

```diff
@@ -0,0 +1,973 @@
+                .latest_operation(public_key, log_id)
+                .await
+                .map_err(LogSyncError::LogStore)?;
+
+            let log_height = match latest_operation {
+                Some((header, _)) => header.seq_num,
+                // If we don't have this log then continue onto the next without
+                // sending any messages.
+                None => continue,
+            };
+
+            // Calculate from which seq num in the log the remote needs operations.
+            let remote_needs_from = match remote_log_heights_map.get(public_key) {
+                Some(log_heights) => {
+                    match log_heights.iter().find(|(id, _)| *id == *log_id) {
+                        // The log is known by the remote, take their log height
+                        // and plus one.
+                        Some((_, log_height)) => log_height + 1,
+                        // The log is not known, they need from seq num 0
+                        None => 0,
+                    }
+                }
+                // The author is not known, they need from seq num 0.
+                None => 0,
+            };
+
+            if remote_needs_from <= log_height {
+                let log = store
+                    .get_raw_log(public_key, log_id, Some(remote_needs_from))
+                    .await
+                    .map_err(LogSyncError::LogStore)?;
+
+                if let Some(log) = log {
+                    operations.extend(log);
+                }
+            };
+        }
+    }
+
+    Ok(operations)
+}
+
+/// Protocol messages.
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
+#[serde(tag = "type", content = "value")]
+pub enum LogSyncMessage<L> {
+    Have(Vec<(PublicKey, Vec<(L, u64)>)>),
+    PreSync {
+        total_operations: u64,
+        total_bytes: u64,
+    },
+    Operation(Vec<u8>, Option<Vec<u8>>),
+    Done,
+}
+
+/// Events emitted from log sync sessions.
+#[derive(Clone, Debug, PartialEq)]
+pub enum LogSyncEvent<E> {
+    Status(StatusEvent),
+    Data(Operation<E>),
+}
+
+/// Sync session metrics emitted in session event messages.
+#[derive(Clone, Debug, PartialEq, Default)]
+pub struct Metrics {
+    pub total_operations_local: Option<u64>,
+    pub total_operations_remote: Option<u64>,
+    pub total_bytes_local: Option<u64>,
+    pub total_bytes_remote: Option<u64>,
+}
+
+/// Sync session status variants carried on log sync session events.
+#[derive(Clone, Debug, PartialEq)]
+pub enum StatusEvent {
+    Started {
+        metrics: Metrics,
+    },
+    Progress {
+        metrics: Metrics,
+    },
+    Completed {
+        metrics: Metrics,
+    },
+    Failed {
+        error_message: String,
+        metrics: Metrics,
+    },
+}
+
+/// Operations carried on log sync session events.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Operation<E> {
+    pub(crate) header: Header<E>,
+    pub(crate) body: Option<Body>,
+}
```

Comment:
Is it possible to use `Operation` coming from core? It will be confusing to introduce another, independent type with the same name (and we should in general avoid introducing too many different representations of the same thing, that at least was coming from feedback)